mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -402,4 +445,402 @@ class BlockMatrix @Since("1.3.0") (
+  }
+
+
+
+  /** Computes the LU Decomposition of a Square Matrix.  For a matrix A of size (n x n)
+    * LU decomposition computes the Lower Triangular Matrix L, the Upper Triangular
+    * Matrix U, along with a Permutation Matrix P, such that PA=LU.  The Permutation
+    * Matrix addresses cases where zero entries prevent forward substitution
+    * solution of L or U.
+    *
+    * The BlockMatrix version takes a BlockMatrix as an input and returns a Tuple
+    * of 5 BlockMatrix objects:
+    * P, L, U (in that order), such that P.multiply(A)-L.multiply(U) = 0
+    * and Li, Ui, which are the inverse of the block diagonal terms for L and U.
+    *
+    * The blockLU method will return only P,L, and U, but blockLUtoSolver will return
+    * the extra Li and Ui matrices, which will be used by the solve method
+    * so that it does not need to recompute these values.
+    *
+    * The method follows a procedure similar to the method used in ScaLAPACK, but
+    * places more emphasis on preparing BlockMatrix objects as inputs to large
+    * BlockMatrix.multiply operations.
+    *
+    *
+    * @return P,L,U,Li,Ui as a Tuple of BlockMatrix
+    * @since 1.6.0
+  */
+
+  private [mllib] def blockLUtoSolver:
+          (BlockMatrix, BlockMatrix, BlockMatrix, BlockMatrix, BlockMatrix) = {
+
+    // builds up the array as a union of RDD sets
+    val nDiagBlocks = this.numColBlocks
+    // Matrix changes shape during recursion...the "absolute location" must be
+    // preserved when reconstructing.
+    val rowsAbs = this.numRowBlocks; val colsAbs = rowsAbs
+    // accessing the spark context
+    val sc = this.blocks.sparkContext
+
+    /** Recursive Sequence Build is a nested recursion method that builds up all of the
+      * sequences that are converted to BlockMatrix classes for large matrix
+      * multiplication operations.  The Schur Complement is calculated at each
+      * recursion step and fed into the next iteration as the input matrix.
+      *
+      * dP, dL, dU, dLi, dUi have the solutions to LU(S) in the (i,i) (diagonal) blocks.
+      * dLi and dUi, are the inverses of each block in dL and dU.  UD and LD contain the
+      * extracted subBlocks from the incoming matrix (which is the Schur Complement
+      * from the previous iteration).  These matrices occupy the U12 and L21 spaces at
+      * each iteration, and form the strict upper and lower block diagonal matrices,
+      * respectively.  This means that for UD, only (i,j>i) blocks are populated with
+      * the cascading Schur calculations, while for LD, (i, j<i) blocks are populated.
+      *
+      * @param rowI
+      * @param prevTuple
+      * @return dP, dL, dU, dLi, dUi, LD, UD, S  All are RDDs of Sequences that are
+      *         iteratively built, while S is a BlockMatrix used in the recursion loop
+      * @since 1.6.0
+      */
+    def recursiveSequencesBuild(rowI: Int, prevTuple:
+            (RDD[((Int, Int), Matrix)], RDD[((Int, Int), Matrix)],
```

Comment:
Here, I added a (non case) class within the blockLUtoSolver method.  @frress also suggested using case classes in other portions of the code to improve readability.  
Please let me know if this is what you were thinking of...I'd like to continue to clean this up as time permits.